.github/actions/test_gem/action.yml

```diff
@@ -0,0 +1,130 @@
+        # of the benefits of bundler caching.
+        rm -f "${dir}/Gemfile.lock"
+
+        echo "::set-output name=cache_key::mri"
+        if [[ "${{ inputs.ruby }}" == "jruby" ]]; then
+          echo "::set-output name=cache_key::jruby"
+        fi
+
+        echo "::set-output name=appraisals::false"
+        if [[ -f "${dir}/Appraisals" ]]; then
+          echo "::set-output name=appraisals::true"
+        fi
+
+    # Install ruby and bundle dependencies and cache!
+    # ...but not for appraisals, sadly.
+    - name: Install Ruby ${{ inputs.ruby }} with dependencies
+      if: "${{ steps.setup.outputs.appraisals == 'false' }}"
+      uses: ruby/setup-ruby@v1
+      with:
+        ruby-version: "${{ inputs.ruby }}"
+        working-directory: "${{ steps.setup.outputs.dir }}"
+        bundler-cache: true
+        cache-version: "v1-${{ steps.setup.outputs.cache_key }}"
+
+    # If we're using appraisals, do it all manually.
+    - name: Install Ruby ${{ inputs.ruby }} without dependencies
+      if: "${{ steps.setup.outputs.appraisals == 'true' }}"
+      uses: ruby/setup-ruby@v1
+      with:
+        ruby-version: "${{ inputs.ruby }}"
+    - name: Install dependencies and appraisals
+      if: "${{ steps.setup.outputs.appraisals == 'true' }}"
+      shell: bash
+      run: |
+        cd "${{ steps.setup.outputs.dir }}"
+        bundle install
+        bundle exec appraisal install
+
+    - name: Test Gem
+      shell: bash
+      run: |
+        cd "${{ steps.setup.outputs.dir }}"
+
+        if [[ -f "Appraisals" ]]; then
+          bundle exec appraisal rake test
+        else
+          bundle exec rake test
+        fi
+      env:
+        TEST_KAFKA_HOST: "127.0.0.1"
+        TEST_KAFKA_PORT: 29092
+        TEST_MYSQL_HOST: "127.0.0.1"
+        TEST_MYSQL_PORT: 3306
+        TEST_MYSQL_USER: mysql
+        TEST_MYSQL_PASSWORD: mysql
+        TEST_POSTGRES_PASSWORD: postgres
+        TEST_POSTGRES_USER: postgres
+        TEST_POSTGRES_HOST: localhost
+        TEST_POSTGRES_PORT: 5432
+        TEST_POSTGRES_DB: postgres
+        TEST_MEMCACHED_HOST: localhost
+        TEST_MEMCACHED_PORT: 11211
+        TEST_MONGODB_HOST: localhost
+        TEST_MONGODB_PORT: 27017
+        TEST_RABBITMQ_HOST: localhost
+        TEST_RABBITMQ_PORT: 5672
+        TEST_RABBITMQ_URL: amqp://guest:guest@localhost:5672
+        TEST_REDIS_HOST: localhost
+        TEST_REDIS_PORT: 6379
+
+    - name: YARD
+      shell: bash
+      if: "${{ inputs.yard == 'true' }}"
+      run: |
+        cd "${{ steps.setup.outputs.dir }}"
+        bundle exec rake yard
+
+    - name: Rubocop
+      shell: bash
+      if: "${{ inputs.rubocop == 'true' }}"
+      run: |
+        cd "${{ steps.setup.outputs.dir }}"
+        bundle exec rake rubocop
+
+    - name: Build Gem
+      shell: bash
+      if: "${{ inputs.build == 'true' }}"
+      run: |
+        cd "${{ steps.setup.outputs.dir }}"
+        gem build ${{ inputs.gem }}.gemspec
```

Comment:
✨ 

I wonder if installing the gem also will give us an indication of failure if there are un-released incompatible versions?

Or maybe that should be part of the release build?

Reply:
It might? I understand what you're getting at. I wonder if it's possible to actually do so before all the gems are pushed to rubygems? For example, if `opentelemetry-instrumentation-foo` depends on `opentelemetry-sdk` v2, but only `opentelemetry-sdk` v1 is currently released, it would necessarily fail.

That _may_ just be an edge case not worth considering, though (specifically when we are bumping the version of one gem in a PR and bumping the dependencies of another gem in the _same_ PR) ? 

I'm interested in what you think about that. If the barriers don't seem insurmountable, we could do that fairly easily in a follow-up PR to this one. Or in this one, depending on how my waning motivation for the day holds up. 😆

Reply:
It would make more sense as a follow up PR. It is unrelated to this problem and this PR got my mental juices flowing.

Reply:
It's absolutely a good idea though, it's another way to stop our release 🦶 🔫  problems.